### PR TITLE
update cookbook to chef 12.5 custom resource

### LINKED ISSUE #54: FC059: declare use_inline_resources
Use `use_inline_resources` as per [FC059](http://www.foodcritic.io/#FC059) or resolve the issue some other way. We ended up reverting this for v0.7.4 and created this ticket to address it later as it is currently breaking folks. 

See #51 for more details.

### LINKED ISSUE #60: sysctl::apply always updates 2 resources
is there a way to prevent this from updating 2 resources, on every chef run, from the notifies that happen with the apply recipe? we expect chef runs to always have 0 resources updated, until an actual change has happened.

### PR DESCRIPTION
**Pull request by chasebolt**
this is a big refactor here. i kept backwards compatibility as much as possible, but it would still be best to bump the major version of this cookbook.

this can be 100% library driven cookbook, based off of how i maintain the docker cookbook. you can continue using `sysctl_param` style, but no longer required to include the sysctl recipes; it will just work.

you can also continue to use the attribute method, but it will require adding `include_recipe sysctl` in your run list. i have kept the recipe `sysctl::apply` for backwards compatibility; it just includes the default recipe.

resolves #60
resolves #54

### CODE REVIEW

**Comment by svanzoest** [comment]
Hi @chasebolt,

Thanks for your work on this. I general I would have taken a very similar approach. I am going to find some time to look this over and see how it may impact current installations and look towards merging.

I am primarily going to focus on platform specific behavior as some of this makes it more generic, which maybe okay, but I want to be sure. Also the procps init file came directly from the OS vendor itself, so I think limiting it to just `*.conf` likely breaks backwards compatibility.

Anyways..I wanted you to know that your work on this is appreciated and need to find time to test and verify.

**Comment by chasebolt** [comment]
No worries! This supports a large number of distros and hopefully I covered
each case correctly. I am currently running this in production for centos
6.

The procps file we can easily just remove the *.conf. I only added it to be
more specific and I can be swayed either way on this. Though I think this
naming is different between distros so we have to decide what distro is
"correct" :)

**Comment by zmingxie** [comment]
@svanzoest Did you find any issues from the review and testing? I cherry-pick @chasebolt patches, and they work great with my CentOS boxes.

**Comment by cakuzo** [comment]
@svanzoest did you see any problems with the changes from chasebolt already?

**Comment by svanzoest** [comment]
@cakuzo @zmingxie I have not completed testing on this, but if you have done tests, can you please let me know what platforms you have tried and what the results were? That will help speed up the process.

**Comment by sbadia** [comment]
hello here, @svanzoest Ok on Ubuntu 16.04 with chef (from opscode) `Chef: 12.16.42`

File: recipes/default.rb [hunk 1]
```diff
@@ -19,58 +19,11 @@
...
-    source 'sysctl.conf.erb'
-    mode '0644'
-    notifies :restart, 'service[procps]', :immediately
+coerce_attributes(node['sysctl']['params']).each do |x|
+  k, v = x.split('=')
+  sysctl_param k do
+    value v
   end
-end
+end if node['sysctl']['params']
```
Comment (sbadia) [inline recipes/default.rb]: On a bare node, with the recipe in the `run_list` but without any sysctl params, this fail with NoMethodError `undefined method '[]' for nil:NilClass`

We can change this condition to something like that :)

```ruby
end if node['sysctl']
```

File: recipes/default.rb [hunk 2]
```diff
@@ -26,4 +26,4 @@
   sysctl_param k do
     value v
   end
-end if node['sysctl']['params']
+end if node['sysctl'].attribute?('params')
```
Comment (sbadia) [inline recipes/default.rb]: Hi, thanks for this update and your reactivity!! But the problem is still present (the problem occur when the node definition doesn't contain any sysctl configs (normal/node/…))

```chef
================================================================================
Recipe Compile Error in /var/chef/cache/cookbooks/base-system/recipes/default.rb
================================================================================

NoMethodError
-------------
undefined method `attribute?' for nil:NilClass

Cookbook Trace:
---------------
  /var/chef/cache/cookbooks/sysctl/recipes/default.rb:29:in `from_file'
  /var/chef/cache/cookbooks/base-system/recipes/default.rb:29:in `from_file'

Relevant File Content:
----------------------
/var/chef/cache/cookbooks/sysctl/recipes/default.rb:

 22:  ::Chef::Recipe.send(:include, SysctlCookbook::SysctlHelpers::Param)
 23:  
 24:  coerce_attributes(node['sysctl']['params']).each do |x|
 25:    k, v = x.split('=')
 26:    sysctl_param k do
 27:      value v
 28:    end
 29>> end if node['sysctl'].attribute?('params')
 30:  

Platform:
---------
x86_64-linux
```

**Comment by chasebolt** [comment]
this is a pretty old PR. is there an interest in getting this merged?

**Comment by damacus** [comment]
Yeah it's terribly old, apologies for that. 

So old it needs rebasing or we splat.

**Comment by chasebolt** [comment]
no worries. ill go through the work of rebasing it when i get some time.

**Comment by chasebolt** [comment]
will check later why travis is failing. kitchen locally passes all tests.

**Comment by chasebolt** [comment]
@damacus all tests have passed

**Comment by damacus** [comment]
Thanks @chasebolt !

**Pull request merged**